### ISSUE TITLE
Reorder any into we

### ISSUE THREAD

**Issue by affranchi** [oneapi-src/oneDNN#786]
Hi,

I'm benchmarking oneDNN LSTM and found that it doesn't need to use tag::any for weight and iteration weight if I'm using tag::abcde for weight desc and iter_weight desc. I tested channel sizes from 128 to 1024 and multi batches up to 5 batches and time steps from 1 to 120. Also, I tested bi-directional case, too.

What I expect is that LSTM shows shorter execution time and primitive descriptor creation time like Convolution primitive. But mostly, LSTM execution time is not that bad except using long time steps. Mostly, memory reorder from abcde to recommended memory (weights_desc and weights_iter_desc) takes some time which makes it slower than the case without using "tag::any" for weight descriptors.

I'm using 1.6-rc now and still get same benchmark result. Is it expected behavior? If so, is there any other tip to reduce latency for LSTM primitive?
FYI, I have run warm-up LSTM before benchmarking for all cases.

Thanks.

**Comment by emfomenk** [comment]
Hi @affranchi,

Could you please specify the hardware you are using, whether this is an inference or training mode, and what data type is used? Having a complete reproducer would be a nice thing.

The thing is that with format tag `any` the library (unfortunately) doesn't know what format you are currently using, hence cannot estimate if the reorder + primitive with desired format is faster than primitive with the current format. This is a limitation of the current API.

With what you described it sounds like:
- the GEMM that we use internally is well optimized for the format `abcde`, so that it is not significantly slower than GEMM with the format that LSTM asked you. And / or
- the LSTM is so small (channels ~= 128, batch <= 5), that GEMM contribution doesn't dominate over the overhead that is caused by a reorder. 
    - You partially confirmed that by saying that you see the gain when there are many timestamps (in general, the more timestamps, the less the contribution of the weights reorder). 
    - I would also expect better performance with format tag `any` when the number of channels is big, say 1024.

**Comment by affranchi** [comment]
Thank you for the answers, @emfomenk 
It's very helpful information.

I'm currently testing over the [oneDNN LSTM example](https://oneapi-src.github.io/oneDNN/lstm_example_cpp.html) by changing parameters.

My testing machine is [AWS M5](https://aws.amazon.com/ec2/instance-types/m5/) which says that it uses "the 1st or 2nd generation Intel Xeon Platinum 8000 series processor (Skylake-SP or Cascade Lake) with a sustained all core Turbo CPU clock speed of up to 3.1 GHz".

I'm still testing to figure out when I need to use tag::any and when I need to not use tag::any.

Would be nice if I can get more insight from you.

Thanks!

**Comment by emfomenk** [comment]
Hi @affranchi,

Did you change an example to run the iterations at least several times? The first run typically doesn't show the representative performance results. Needless to say the example itself isn't written in a way it would show good performance, as, for instance, it copies the memory twice and does that in a single thread. 

I changed the example as shown in [1] below and got the following results:

``` bash
# common settings
export cmd="numactl -m0 -N0 ./examples/primitives-lstm-cpp"
export OMP_NUM_THREADS=24
export DNNL_VERBOSE=1
```

| | Using tag `any` | Forcing the tag to `ldigo` |
| :-- | :-- | :-- |
| Run-line | `ANY=1 $cmd` | `ANY=0 $cmd` |
| LSTM time [*] | 0.58 ms | 0.68 ms |
| Reorders time [*] | 0.02 + 0.04 ms | - |
| Total time | 0.64 ms | 0.68 ms |

[*] -- from the last (10th) run 

So, I would suggest you to try LSTM in the real application, where you want it to apply, and see the difference there. As I said, the example isn't really a good benchmark.

It would be nice if you are able to implement the code in a generic way, and fine-tune / test the overall performance once everything is ready by switching the weight tags (just as in the patch below). You are welcome to share the results if you still see that `any` gives you worse performance. As mentioned in the previous post, that could be a limitation of the API, but could be a bug either, so your feedback would be very helpful (especially with the verbose log).

---
[1]

```patch
diff --git a/examples/primitives/lstm.cpp b/examples/primitives/lstm.cpp
index 2f6a7d9fc..e1274e7ec 100644
--- a/examples/primitives/lstm.cpp
+++ b/examples/primitives/lstm.cpp
@@ -45,7 +45,7 @@ using namespace dnnl;
 using tag = memory::format_tag;
 using dt = memory::data_type;

-void lstm_example(dnnl::engine::kind engine_kind) {
+void lstm_example_exec(dnnl::engine::kind engine_kind) {

     // Create execution dnnl::engine.
     dnnl::engine engine(engine_kind, 0);
@@ -54,11 +54,11 @@ void lstm_example(dnnl::engine::kind engine_kind) {
     dnnl::stream engine_stream(engine);

     // Tensor dimensions.
-    const memory::dim N = 26, // batch size
-            T = 6, // time steps
-            C = 12, // channels
+    const memory::dim N = 5, // batch size
+            T = 20, // time steps
+            C = 128, // channels
             G = 4, // gates
-            L = 4, // layers
+            L = 1, // layers
             D = 1; // directions

     // Source (src), weights, bias, and destination (dst) tensors
@@ -111,10 +111,14 @@ void lstm_example(dnnl::engine::kind engine_kind) {
     write_to_dnnl_memory(weights_layer_data.data(), user_weights_layer_mem);
     write_to_dnnl_memory(weights_iter_data.data(), user_weights_iter_mem);

+    const char *env = getenv("ANY");
+    bool do_not_use_any = env && *env == '0';
+    auto wei_tag = do_not_use_any ? tag::abcde : tag::any;
+
     // Create memory descriptors for weights with format_tag::any. This enables
     // the LSTM primitive to choose the optimized memory layout.
-    auto lstm_weights_layer_md = memory::desc(weights_dims, dt::f32, tag::any);
-    auto lstm_weights_iter_md = memory::desc(weights_dims, dt::f32, tag::any);
+    auto lstm_weights_layer_md = memory::desc(weights_dims, dt::f32, wei_tag);
+    auto lstm_weights_iter_md = memory::desc(weights_dims, dt::f32, wei_tag);

     // Optional memory descriptors for recurrent data.
     auto src_iter_md = memory::desc();
@@ -191,6 +195,10 @@ void lstm_example(dnnl::engine::kind engine_kind) {
     read_from_dnnl_memory(dst_layer_data.data(), dst_layer_mem);
 }

+void lstm_example(dnnl::engine::kind engine_kind) {
+    for (int i = 0; i < 10; ++i) lstm_example_exec(engine_kind);
+}
+
 int main(int argc, char **argv) {
     return handle_example_errors(lstm_example, parse_engine_kind(argc, argv));
 }
```

**Comment by affranchi** [comment]
Thank you so much for your detail answers, @emfomenk !

I close the question for now. I'll be back with more detail result and reproducible implementation.
One of weird observation is that batch size 1 with any tag has worse performance than batch size 10 with any tag when the channel size is 1024 and 1 time stamp case.

BatchSize | Any result(us) | No any result(us)
1                | 5407                | 785
10              | 1972                 | 1496

So far, I observe this thing consistently but looks it could be affected by which test is done first. (I'm not so sure of this part that's why I'll come back with more verification)

FYI, it's 1.6-rc version and AWS m5 which is specified in the above.

Thanks.

**Comment by emfomenk** [comment]
Thanks @affranchi,

> One of weird observation is that batch size 1 with any tag has worse performance than batch size 10 with any tag when the channel size is 1024 and 1 time stamp case.

Yeah, for such a small problem, it seems that reorders take significantly more than the LSTM itself.
Also, for whatever reason the LSTM itself is slightly slower with an _optimized_ layout than with the fixed one.
Here is what I have:

  | Using tag any | Forcing the tag to ldigo
-- | -- | --
Run-line | `ANY=1 $cmd` | `ANY=0 $cmd`
LSTM time [*] | 0.79 ms | 0.69 ms
Reorders time [*] | 0.87 + 1.00 ms | -
Total time | 2.66 ms | 0.69 ms

This could be considered as a performance bug. For the very least one would expect that LSTM time with format tag `any` won't be bigger than with a fixed format tag, like `ldigo`. On the other hand, I am not sure how useful this particular case is (T=1, Batch=1) given the time it consumes (like contribution for the overall performance) + the difference between LSTM. So, maybe we can live with this performance bug for the time being 😅  Please let us know if you think this case is important for the end-user application, so that we can prioritize accordingly.

> I close the question for now. I'll be back with more detail result and reproducible implementation.

Sure! Will be glad to receive feedback!
